crates/accelerate/src/convert_2q_block_matrix.rs

```diff
@@ -53,94 +52,147 @@ pub fn get_matrix_from_inst(py: Python, inst: &PackedInstruction) -> PyResult<Ar
-                        Some(current) => {
-                            let temp = one_qubits_combined.dot(&current);
-                            op_matrix.dot(&temp)
-                        }
-                    });
-                    qubit_0 = ONE_QUBIT_IDENTITY;
-                    qubit_1 = ONE_QUBIT_IDENTITY;
-                    one_qubit_components_modified = false;
-                } else {
-                    output_matrix = Some(match output_matrix {
-                        None => op_matrix,
-                        Some(current) => op_matrix.dot(&current),
-                    });
+            Qarg::Q01 | Qarg::Q10 => {
+                let mut matrix = get_matrix_from_inst(py, inst)?;
+                if qarg == Qarg::Q10 {
+                    change_basis_inplace(matrix.view_mut());
                 }
-            }
-            [1, 0] => {
-                let matrix = change_basis(op_matrix.view());
-                if one_qubit_components_modified {
-                    let one_qubits_combined = kron(&aview2(&qubit_1), &aview2(&qubit_0));
-                    output_matrix = Some(match output_matrix {
-                        None => matrix.dot(&one_qubits_combined),
-                        Some(current) => matrix.dot(&one_qubits_combined.dot(&current)),
-                    });
-                    qubit_0 = ONE_QUBIT_IDENTITY;
-                    qubit_1 = ONE_QUBIT_IDENTITY;
-                    one_qubit_components_modified = false;
-                } else {
-                    output_matrix = Some(match output_matrix {
-                        None => matrix,
-                        Some(current) => matrix.dot(&current),
-                    });
+                if let Some(sep) = qubits_1q.take() {
+                    matrix = matrix.dot(&sep.matrix_into(&mut work));
                 }
+                output_matrix = if let Some(state) = output_matrix {
+                    Some(matrix.dot(&state))
+                } else {
+                    Some(matrix)
+                };
             }
-            _ => unreachable!(),
         }
     }
-    Ok(match output_matrix {
-        Some(matrix) => {
-            if one_qubit_components_modified {
-                let one_qubits_combined = kron(&aview2(&qubit_1), &aview2(&qubit_0));
-                one_qubits_combined.dot(&matrix)
-            } else {
-                matrix
-            }
-        }
-        None => kron(&aview2(&qubit_1), &aview2(&qubit_0)),
-    })
+    match (
+        qubits_1q.map(|sep| sep.matrix_into(&mut work)),
+        output_matrix,
+    ) {
+        (Some(sep), Some(state)) => Ok(sep.dot(&state)),
+        (None, Some(state)) => Ok(state),
+        (Some(sep), None) => Ok(sep.to_owned()),
+        // This shouldn't actually ever trigger, because we expect blocks to be non-empty, but it's
+        // trivial to handle anyway.
+        (None, None) => Ok(arr2(&TWO_QUBIT_IDENTITY)),
+    }
```

Comment:
Not for this PR, but it'll be great to have this all done via `Matrix4` after this merges, so we can just consolidate to a `UnitaryGate` with `ArrayType::TwoQ` and avoid allocating a matrix altogether.

Reply:
Yeah absolutely - I stopped in this PR at the point that I started interacting with the 2q matrices because I wasn't prepared to put too much more time into things I knew were both in flux and needed more major changes.